### PR TITLE
near-vm-logic: test log methods behaviour when memory access is violated

### PR DESCRIPTION
**Pull request by mina86**

### CODE REVIEW

**Comment by nagisa** [comment]
Could you please split the change into two commits where one of them applies changes you were making and the other moves the code around? As-is it is pretty difficult to review the code that has moved for absence of changes.

**Comment by mina86** [comment]
Done.

File: runtime/near-vm-logic/src/tests/miscs.rs [hunk 1]
```diff
@@ -398,6 +398,121 @@ fn test_valid_log_utf16_null_terminated_fail() {
...
+        let bytes = b"foo bar \xff baz qux";
+        let bytes = logic.internal_mem_write_at(64 * 1024 - bytes.len() as u64, bytes);
+        let err = if read_ok { HostError::BadUTF8 } else { HostError::MemoryAccessViolation };
+        assert_eq!(Err(err.into()), test(&mut logic, bytes));
+    }
+
+    #[test]
+    fn test_good_read() {
+        // The data is read correctly but it has invalid UTF-8 thus it ends up
+        // with BadeUTF8 error and user being charged for decoding.
```
Comment (jakmeier) [inline runtime/near-vm-logic/src/tests/miscs.rs]:
```suggestion
        // with BadUTF8 error and user being charged for decoding.
```

File: runtime/near-vm-logic/src/tests/miscs.rs [hunk 2]
```diff
@@ -398,6 +398,121 @@ fn test_valid_log_utf16_null_terminated_fail() {
...
+        append_utf16(&mut bytes, "foobarbaz");
+        let bytes = logic.internal_mem_write_at(64 * 1024 - bytes.len() as u64, &bytes);
+        let err = if read_ok { HostError::BadUTF16 } else { HostError::MemoryAccessViolation };
+        assert_eq!(Err(err.into()), test(&mut logic, bytes));
+    }
+
+    #[test]
+    fn test_good_read() {
+        // The data is read correctly but it has invalid UTF-16 thus it ends up
+        // with BadeUTF16 error and user being charged for decoding.
```
Comment (jakmeier) [inline runtime/near-vm-logic/src/tests/miscs.rs]:
```suggestion
        // with BadUTF16 error and user being charged for decoding.
```

**Review by jakmeier — APPROVED** [review]
LGTM, zhank you for the additional test coverage!

File: runtime/near-vm-logic/src/tests/logs.rs [hunk 3]
```diff
@@ -0,0 +1,510 @@
...
+        let bytes = b"foo bar \xff baz qux";
+        let bytes = logic.internal_mem_write_at(64 * 1024 - bytes.len() as u64, bytes);
+        let err = if read_ok { HostError::BadUTF8 } else { HostError::MemoryAccessViolation };
+        assert_eq!(Err(err.into()), test(&mut logic, bytes));
+    }
+
+    #[test]
+    fn test_good_read() {
+        // The data is read correctly but it has invalid UTF-8 thus it ends up
+        // with BadeUTF8 error and user being charged for decoding.
```
Comment (jakmeier) [inline runtime/near-vm-logic/src/tests/logs.rs:413]:
```suggestion
        // with BadUTF8 error and user being charged for decoding.
```

File: runtime/near-vm-logic/src/tests/logs.rs [hunk 4]
```diff
@@ -0,0 +1,510 @@
...
+        append_utf16(&mut bytes, "foobarbaz");
+        let bytes = logic.internal_mem_write_at(64 * 1024 - bytes.len() as u64, &bytes);
+        let err = if read_ok { HostError::BadUTF16 } else { HostError::MemoryAccessViolation };
+        assert_eq!(Err(err.into()), test(&mut logic, bytes));
+    }
+
+    #[test]
+    fn test_good_read() {
+        // The data is read correctly but it has invalid UTF-16 thus it ends up
+        // with BadeUTF16 error and user being charged for decoding.
```
Comment (jakmeier) [inline runtime/near-vm-logic/src/tests/logs.rs:472]:
```suggestion
        // with BadUTF16 error and user being charged for decoding.
```

**Pull request merged**